# ðŸ¤– Prompt per Jules - SonarCloud Fix

**Generato:** 2025-12-07 19:28
**Progetto:** gianky00_formazione_coemi

---

## Come usare questi prompt

1. Vai su [jules.google.com](https://jules.google.com/)
2. Connetti il tuo repository GitHub
3. Copia il prompt desiderato
4. Allega il file Markdown corrispondente (RELIABILITY, COMPACT, o ULTRA_COMPLETE)
5. Avvia Jules e lascialo lavorare!

---

## ðŸŽ¯ Prompt Principale (Completo)

# ðŸŽ¯ Prompt per Jules - Fix Completo SonarCloud

## Contesto
Devi correggere 665 problemi di codice identificati da SonarCloud nel repository.
I problemi sono distribuiti in 115 file.
Tempo stimato per completare tutti i fix: **7g 6h 2min**

## Statistiche
- ðŸ”´ **RELIABILITY**: 42 issues (~3h 30min)
- ðŸŸ¡ **MAINTAINABILITY**: 623 issues (~7g 2h 32min)

## Problemi piÃ¹ frequenti
- Functions and methods should not be empty (129x)
- Field names should comply with a naming convention (94x)
- Unused local variables should be removed (62x)
- React components should validate prop types (54x)
- Cognitive Complexity of functions should not be too high (46x)

## Istruzioni
Per ogni issue nel documento allegato:
1. **Apri il file** indicato nel path
2. **Vai alla riga** specificata
3. **Leggi il problema** nella sezione "ðŸŽ¯ Problema"
4. **Applica la correzione** seguendo la sezione "âœ… Come risolvere"
5. **Verifica** che il codice compili/funzioni dopo la modifica

## PrioritÃ  di correzione
2. ðŸ”´ **RELIABILITY** - Bug (crash/errori) - ALTO
3. ðŸŸ¡ **MAINTAINABILITY** - Code smell - MEDIO

## Note importanti
- Se un fix richiede modifiche a piÃ¹ file, assicurati di mantenere la coerenza
- Per i fix complessi, aggiungi commenti esplicativi nel codice
- Se non sei sicuro di un fix, salta quell'issue e passa al successivo
- Committa i cambiamenti in modo incrementale (es. per tipo di problema)

## Comando suggerito per il commit
```
git add -A && git commit -m "fix: risolti {total} issues SonarCloud ({', '.join(f'{issues_count.get(q, 0)} {q}' for q in SOFTWARE_QUALITIES if issues_count.get(q, 0) > 0)})"
```


---

## ðŸ”´ Prompt RELIABILITY

# ðŸ”´ Prompt per Jules - Fix RELIABILITY

## Obiettivo
Correggi tutti i **42 problemi di RELIABILITY** nel repository.
Tempo stimato: **3h 30min**

## Cos'Ã¨ RELIABILITY?
Bug che possono causare crash o comportamenti errati

## File piÃ¹ problematici
- `guide_frontend/src/components/DashboardSimulator.jsx` (10 issues)
- `guide_frontend/src/components/ui/Accordion.jsx` (6 issues)
- `guide_frontend/src/components/ui/GuideCard.jsx` (5 issues)
- `guide_frontend/src/components/Sidebar.jsx` (5 issues)
- `guide_frontend/src/components/ui/Badge.jsx` (3 issues)

## Tipi di problemi piÃ¹ frequenti
- **React components should validate prop types** (27x)
- **Non-interactive DOM elements should not have an interactive handler** (5x)
- **Mouse events should have corresponding keyboard events** (5x)
- **Spacing between inline elements should be explicit** (2x)
- **Floating point numbers should not be tested for equality** (1x)

## Istruzioni specifiche per RELIABILITY
1. Apri il file `JULES_RELIABILITY_ONLY_*.md` allegato
2. Per ogni issue:
   - Vai al file e riga indicati
   - Leggi il codice attuale (sezione "ðŸ’» Codice attuale")
   - Applica il fix suggerito (sezione "âœ… Come risolvere")
   - Verifica che non ci siano regressioni

## Suggerimento
Per i fix ripetitivi dello stesso tipo, puoi usare "Trova e sostituisci" nell'IDE
o creare uno script di refactoring automatico.

## Commit suggerito
```
git add -A && git commit -m "fix(reliability): risolti 42 issues RELIABILITY"
```


---

## ðŸŸ¡ Prompt MAINTAINABILITY

# ðŸŸ¡ Prompt per Jules - Fix MAINTAINABILITY

## Obiettivo
Correggi tutti i **623 problemi di MAINTAINABILITY** nel repository.
Tempo stimato: **7g 2h 32min**

## Cos'Ã¨ MAINTAINABILITY?
Code smell che rendono il codice difficile da mantenere

## File piÃ¹ problematici
- `tests/desktop_app/mock_qt.py` (259 issues)
- `app/core/config.py` (18 issues)
- `launcher.py` (15 issues)
- `tests/app/core/test_db_security_integration.py` (14 issues)
- `desktop_app/main_window_ui.py` (14 issues)

## Tipi di problemi piÃ¹ frequenti
- **Functions and methods should not be empty** (129x)
- **Field names should comply with a naming convention** (94x)
- **Unused local variables should be removed** (62x)
- **Cognitive Complexity of functions should not be too high** (46x)
- **"SystemExit" should be re-raised** (37x)

## Istruzioni specifiche per MAINTAINABILITY
1. Apri il file `JULES_MAINTAINABILITY_ONLY_*.md` allegato
2. Per ogni issue:
   - Vai al file e riga indicati
   - Leggi il codice attuale (sezione "ðŸ’» Codice attuale")
   - Applica il fix suggerito (sezione "âœ… Come risolvere")
   - Verifica che non ci siano regressioni

## Suggerimento
Per i fix ripetitivi dello stesso tipo, puoi usare "Trova e sostituisci" nell'IDE
o creare uno script di refactoring automatico.

## Commit suggerito
```
git add -A && git commit -m "fix(maintainability): risolti 623 issues MAINTAINABILITY"
```


---

## âš¡ Prompt Rapido

# âš¡ Prompt Rapido per Jules

Correggi 665 issues SonarCloud in 115 file.

**PrioritÃ :**
ðŸ”´ RELIABILITY: 42
ðŸŸ¡ MAINTAINABILITY: 623

**Tempo stimato:** 7g 6h 2min

**Cosa fare:**
1. Apri il file JULES_ULTRA_COMPLETE_*.md
2. Per ogni sezione, vai al file:riga indicati
3. Applica il fix nella sezione "Come risolvere"
4. Testa e committa


---

## ðŸ”„ Prompt per Fix in Batch

# ðŸ”„ Prompt per Fix in Batch (per tipo di problema)

Alcuni problemi si ripetono molte volte. Ecco come risolverli in batch:

## Top 10 problemi piÃ¹ frequenti

### Functions and methods should not be empty
- **Occorrenze:** 129
- **Regola:** `python:S1186`
- **Fix rapido:** ##### Noncompliant code example

```
def shouldNotBeEmpty(): # Noncompliant - method is empty
 pass

def notImplemented(): # Noncompliant - method is empty
 pass

def emptyOnPurpose(): # Noncompliant ...

### Field names should comply with a naming convention
- **Occorrenze:** 94
- **Regola:** `python:S116`
- **Fix rapido:** 

### Unused local variables should be removed
- **Occorrenze:** 62
- **Regola:** `python:S1481`
- **Fix rapido:** The fix for this issue is straightforward. Once you ensure the unused variable is not part of an incomplete implementation leading to bugs, you
just need to remove it.

##### Noncompliant code example...

### React components should validate prop types
- **Occorrenze:** 54
- **Regola:** `javascript:S6774`
- **Fix rapido:** ##### Noncompliant code example

```
import PropTypes from 'prop-types';

function Hello({ firstname, lastname }) {
 return &lt;div&gt;Hello {firstname} {lastname}&lt;/div&gt;; // Noncompliant: 'lastn...

### Cognitive Complexity of functions should not be too high
- **Occorrenze:** 46
- **Regola:** `python:S3776`
- **Fix rapido:** Reducing cognitive complexity can be challenging.
 Here are a few suggestions:

 - **Extract complex conditions in a new function.**
 Mixed operators in condition will increase complexity. Extracting ...

### "SystemExit" should be re-raised
- **Occorrenze:** 37
- **Regola:** `python:S5754`
- **Fix rapido:** Re-raise `SystemExit`, `BaseException` and any exceptions caught in a bare `except` clause.

##### Noncompliant code example

```
try:
 ...
except SystemExit: # Noncompliant: the SystemExit exception ...

### Method names should comply with a naming convention
- **Occorrenze:** 30
- **Regola:** `python:S100`
- **Fix rapido:** 

### String literals should not be duplicated
- **Occorrenze:** 29
- **Regola:** `python:S1192`
- **Fix rapido:** Use constants to replace the duplicated string literals. Constants can be referenced from many places, but only need to be updated in a single
place.

##### Noncompliant code example

With the default...

### Unused function parameters should be removed
- **Occorrenze:** 24
- **Regola:** `python:S1172`
- **Fix rapido:** Having unused function parameters in your code can lead to confusion and misunderstanding of a developerâ€™s intention. They reduce code readability
and introduce the potential for errors. To avoid thes...

### Unnecessary imports should be removed
- **Occorrenze:** 22
- **Regola:** `javascript:S1128`
- **Fix rapido:** 


---
