# ðŸ“Š SonarCloud Analysis Summary

**Progetto:** gianky00_formazione_coemi
**Data analisi:** 2025-12-07 22:13
**Generato da:** SonarCloud Exporter v3.0

---

## ðŸ“ˆ Statistiche Generali

| Metrica | Valore |
|---------|--------|
| Totale issues | **150** |
| File coinvolti | 51 |
| Tempo stimato fix | **2g 6h 13min** |
| Regole violate | 25 |

## ðŸŽ¯ Issues per Categoria

| Categoria | Issues | Tempo stimato | PrioritÃ  |
|-----------|--------|---------------|----------|
| ðŸŸ¡ MAINTAINABILITY | 150 | 2g 6h 13min | ðŸŸ¡ MEDIA |

## âš ï¸ Issues per SeveritÃ 

| SeveritÃ  | Issues |
|----------|--------|
| ðŸ”´ CRITICAL | 89 |
| ðŸŸ¡ MAJOR | 34 |
| ðŸŸ¢ MINOR | 27 |

## ðŸ† Top 10 Regole piÃ¹ Violate

| # | Regola | Occorrenze | Descrizione |
|---|--------|------------|-------------|
| 1 | `python:S3776` | 46 | Cognitive Complexity of functions should not be to |
| 2 | `python:S1192` | 23 | String literals should not be duplicated |
| 3 | `python:S5754` | 16 | "SystemExit" should be re-raised |
| 4 | `python:S1481` | 11 | Unused local variables should be removed |
| 5 | `python:S6711` | 8 | numpy.random.Generator should be preferred to nump |
| 6 | `python:S6397` | 6 | Character classes in regular expressions should no |
| 7 | `python:S125` | 5 | Sections of code should not be commented out |
| 8 | `python:S2772` | 5 | "pass" should not be used needlessly |
| 9 | `python:S3457` | 5 | String formatting should be used correctly |
| 10 | `python:S1066` | 3 | Mergeable "if" statements should be combined |

## ðŸ’» Linguaggi

| Linguaggio | Issues |
|------------|--------|
| python | 150 |

## ðŸ“ File Generati

- **JULES_COMPACT_20251207_2213.md** (16.9 KB)
  - 150 issues in formato compatto
- **JULES_ULTRA_COMPLETE_20251207_2213.md** (406.2 KB)
  - 150 issues con tutte le informazioni

---

## ðŸ¤– Prompt per Jules

Copia-incolla uno dei seguenti prompt quando usi Jules:

### Prompt Principale (Completo)

```
# ðŸŽ¯ Prompt per Jules - Fix Completo SonarCloud

## Contesto
Devi correggere 150 problemi di codice identificati da SonarCloud nel repository.
I problemi sono distribuiti in 51 file.
Tempo stimato per completare tutti i fix: **2g 6h 13min**

## Statistiche
- ðŸŸ¡ **MAINTAINABILITY**: 150 issues (~2g 6h 13min)

## Problemi piÃ¹ frequenti
- Cognitive Complexity of functions should not be too high (46x)
- String literals should not be duplicated (23x)
- "SystemExit" should be re-raised (16x)
- Unused local variables should be removed (11x)
- numpy.random.Generator should be preferred to numpy.random.RandomState (8x)

## Istruzioni
Per ogni issue nel documento allegato:
1. **Apri il file** indicato nel path
2. **Vai alla riga** specificata
3. **Leggi il problema** nella sezione "ðŸŽ¯ Problema"
4. **Applica la correzione** seguendo la sezione "âœ… Come risolvere"
5. **Verifica** che il codice compili/funzioni dopo la modifica

## PrioritÃ  di correzione
3. ðŸŸ¡ **MAINTAINABILITY** - Code smell - MEDIO

## Note importanti
- Se un fix richiede modifiche a piÃ¹ file, assicurati di mantenere la coerenza
- Per i fix complessi, aggiungi commenti esplicativi nel codice
- Se non sei sicuro di un fix, salta quell'issue e passa al successivo
- Committa i cambiamenti in modo incrementale (es. per tipo di problema)

## Comando suggerito per il commit
```
git add -A && git commit -m "fix: risolti {total} issues SonarCloud ({', '.join(f'{issues_count.get(q, 0)} {q}' for q in SOFTWARE_QUALITIES if issues_count.get(q, 0) > 0)})"
```

```

### Prompt Rapido

```
# âš¡ Prompt Rapido per Jules

Correggi 150 issues SonarCloud in 51 file.

**PrioritÃ :**
ðŸŸ¡ MAINTAINABILITY: 150

**Tempo stimato:** 2g 6h 13min

**Cosa fare:**
1. Apri il file JULES_ULTRA_COMPLETE_*.md
2. Per ogni sezione, vai al file:riga indicati
3. Applica il fix nella sezione "Come risolvere"
4. Testa e committa

```

### Prompt ðŸŸ¡ MAINTAINABILITY

```
# ðŸŸ¡ Prompt per Jules - Fix MAINTAINABILITY

## Obiettivo
Correggi tutti i **150 problemi di MAINTAINABILITY** nel repository.
Tempo stimato: **2g 6h 13min**

## Cos'Ã¨ MAINTAINABILITY?
Code smell che rendono il codice difficile da mantenere

## File piÃ¹ problematici
- `desktop_app/components/neural_3d.py` (12 issues)
- `desktop_app/views/scadenzario_view.py` (10 issues)
- `app/api/main.py` (8 issues)
- `desktop_app/main.py` (7 issues)
- `admin/offusca/build_dist.py` (7 issues)

## Tipi di problemi piÃ¹ frequenti
- **Cognitive Complexity of functions should not be too high** (46x)
- **String literals should not be duplicated** (23x)
- **"SystemExit" should be re-raised** (16x)
- **Unused local variables should be removed** (11x)
- **numpy.random.Generator should be preferred to numpy.random.RandomState** (8x)

## Istruzioni specifiche per MAINTAINABILITY
1. Apri il file `JULES_MAINTAINABILITY_ONLY_*.md` allegato
2. Per ogni issue:
   - Vai al file e riga indicati
   - Leggi il codice attuale (sezione "ðŸ’» Codice attuale")
   - Applica il fix suggerito (sezione "âœ… Come risolvere")
   - Verifica che non ci siano regressioni

## Suggerimento
Per i fix ripetitivi dello stesso tipo, puoi usare "Trova e sostituisci" nell'IDE
o creare uno script di refactoring automatico.

## Commit suggerito
```
git add -A && git commit -m "fix(maintainability): risolti 150 issues MAINTAINABILITY"
```

```

### Prompt per Fix in Batch

```
# ðŸ”„ Prompt per Fix in Batch (per tipo di problema)

Alcuni problemi si ripetono molte volte. Ecco come risolverli in batch:

## Top 10 problemi piÃ¹ frequenti

### Cognitive Complexity of functions should not be too high
- **Occorrenze:** 46
- **Regola:** `python:S3776`
- **Fix rapido:** Reducing cognitive complexity can be challenging.
 Here are a few suggestions:

 - **Extract complex conditions in a new function.**
 Mixed operators in condition will increase complexity. Extracting ...

### String literals should not be duplicated
- **Occorrenze:** 23
- **Regola:** `python:S1192`
- **Fix rapido:** Use constants to replace the duplicated string literals. Constants can be referenced from many places, but only need to be updated in a single
place.

##### Noncompliant code example

With the default...

### "SystemExit" should be re-raised
- **Occorrenze:** 16
- **Regola:** `python:S5754`
- **Fix rapido:** Re-raise `SystemExit`, `BaseException` and any exceptions caught in a bare `except` clause.

##### Noncompliant code example

```
try:
 ...
except SystemExit: # Noncompliant: the SystemExit exception ...

### Unused local variables should be removed
- **Occorrenze:** 11
- **Regola:** `python:S1481`
- **Fix rapido:** The fix for this issue is straightforward. Once you ensure the unused variable is not part of an incomplete implementation leading to bugs, you
just need to remove it.

##### Noncompliant code example...

### numpy.random.Generator should be preferred to numpy.random.RandomState
- **Occorrenze:** 8
- **Regola:** `python:S6711`
- **Fix rapido:** To fix this issue, replace usages of `numpy.random.RandomState` to `numpy.random.Generator`.

##### Noncompliant code example

```
import numpy as np
np.random.seed(42)
x = np.random.randn() # Noncomp...

### Character classes in regular expressions should not contain only one character
- **Occorrenze:** 6
- **Regola:** `python:S6397`
- **Fix rapido:** 

### Sections of code should not be commented out
- **Occorrenze:** 5
- **Regola:** `python:S125`
- **Fix rapido:** 

### "pass" should not be used needlessly
- **Occorrenze:** 5
- **Regola:** `python:S2772`
- **Fix rapido:** 

### String formatting should be used correctly
- **Occorrenze:** 5
- **Regola:** `python:S3457`
- **Fix rapido:** A `printf-`-style format string is a string that contains placeholders, which are replaced by values when the string is printed or
logged. Mismatch in the format specifiers and the arguments provided ...

### Mergeable "if" statements should be combined
- **Occorrenze:** 3
- **Regola:** `python:S1066`
- **Fix rapido:** If merging the conditions seems to result in a more complex code, extracting the condition or part of it in a named function or variable is a
better approach to fix readability.

##### Noncompliant co...

```
